File: CacheManager.cs
using System;
using System.Collections.Generic;
using System.Threading;

namespace Swell.Infrastructure.Caching
{
    /// <summary>
    /// A thread-safe in-memory cache with per-entry TTL and a bounded capacity
    /// using least-recently-used (LRU) eviction.
    /// </summary>
    public sealed class CacheManager<TKey, TValue> where TKey : notnull
    {
        private sealed class CacheEntry
        {
            public TValue Value { get; set; }
            public DateTime ExpiresAt { get; set; }
            public LinkedListNode<TKey> Node { get; set; }
        }

        private readonly int _capacity;
        private readonly TimeSpan _defaultTtl;
        private readonly Dictionary<TKey, CacheEntry> _entries;
        private readonly LinkedList<TKey> _lruOrder;
        private readonly object _sync = new object();

        private long _hits;
        private long _misses;
        private long _evictions;

        public CacheManager(int capacity, TimeSpan defaultTtl)
        {
            if (capacity <= 0)
                throw new ArgumentOutOfRangeException(nameof(capacity));

            _capacity = capacity;
            _defaultTtl = defaultTtl;
            _entries = new Dictionary<TKey, CacheEntry>(capacity);
            _lruOrder = new LinkedList<TKey>();
        }

        public void Set(TKey key, TValue value, TimeSpan? ttl = null)
        {
            lock (_sync)
            {
                var expiresAt = DateTime.UtcNow + (ttl ?? _defaultTtl);

                if (_entries.TryGetValue(key, out var existing))
                {
                    existing.Value = value;
                    existing.ExpiresAt = expiresAt;
                    _lruOrder.Remove(existing.Node);
                    _lruOrder.AddFirst(existing.Node);
                    return;
                }

                if (_entries.Count >= _capacity)
                {
                    EvictLeastRecentlyUsed();
                }

                var node = _lruOrder.AddFirst(key);
                _entries[key] = new CacheEntry
                {
                    Value = value,
                    ExpiresAt = expiresAt,
                    Node = node,
                };
            }
        }

        public bool TryGet(TKey key, out TValue value)
        {
            lock (_sync)
            {
                if (_entries.TryGetValue(key, out var entry))
                {
                    if (entry.ExpiresAt < DateTime.UtcNow)
                    {
                        _entries.Remove(key);
                        _lruOrder.Remove(entry.Node);
                        _misses++;
                        value = default!;
                        return false;
                    }

                    _lruOrder.Remove(entry.Node);
                    _lruOrder.AddFirst(entry.Node);
                    _hits++;
                    value = entry.Value;
                    return true;
                }

                _misses++;
                value = default!;
                return false;
            }
        }

        public TValue GetOrAdd(TKey key, Func<TKey, TValue> factory, TimeSpan? ttl = null)
        {
            if (TryGet(key, out var existing))
            {
                return existing;
            }

            var created = factory(key);
            Set(key, created, ttl);
            return created;
        }

        public bool Remove(TKey key)
        {
            lock (_sync)
            {
                if (_entries.TryGetValue(key, out var entry))
                {
                    _entries.Remove(key);
                    _lruOrder.Remove(entry.Node);
                    return true;
                }
                return false;
            }
        }

        private void EvictLeastRecentlyUsed()
        {
            var lruKey = _lruOrder.Last!.Value;
            _lruOrder.RemoveLast();
            _entries.Remove(lruKey);
            _evictions++;
        }

        public int PurgeExpired()
        {
            var removed = 0;
            var now = DateTime.UtcNow;
            var expiredKeys = new List<TKey>();

            foreach (var kvp in _entries)
            {
                if (kvp.Value.ExpiresAt < now)
                {
                    expiredKeys.Add(kvp.Key);
                }
            }

            foreach (var key in expiredKeys)
            {
                if (_entries.TryGetValue(key, out var entry))
                {
                    _entries.Remove(key);
                    _lruOrder.Remove(entry.Node);
                    removed++;
                }
            }

            return removed;
        }

        public CacheStats GetStats()
        {
            lock (_sync)
            {
                return new CacheStats
                {
                    Count = _entries.Count,
                    Hits = _hits,
                    Misses = _misses,
                    Evictions = _evictions,
                };
            }
        }
    }

    public struct CacheStats
    {
        public int Count { get; set; }
        public long Hits { get; set; }
        public long Misses { get; set; }
        public long Evictions { get; set; }
    }
}

File: test-pr-agent/cache_manager.py
"""In-memory LRU cache with TTL support for the pricing service."""

import time
import threading
from collections import OrderedDict


class CacheEntry:
    def __init__(self, value, ttl_seconds):
        self.value = value
        self.expires_at = time.time() + ttl_seconds

    def is_expired(self):
        return time.time() > self.expires_at


class LruTtlCache:
    """Bounded cache that evicts least-recently-used entries."""

    def __init__(self, max_size=1000, default_ttl=300):
        self._entries = OrderedDict()
        self._max_size = max_size
        self._default_ttl = default_ttl
        self._lock = threading.Lock()
        self.hits = 0
        self.misses = 0

    def get(self, key):
        with self._lock:
            entry = self._entries.get(key)
            if entry is None:
                self.misses += 1
                return None
            if entry.is_expired():
                del self._entries[key]
                self.misses += 1
                return None
            self._entries.move_to_end(key)
            self.hits += 1
            return entry.value

    def put(self, key, value, ttl_seconds=None):
        ttl = ttl_seconds if ttl_seconds else self._default_ttl
        with self._lock:
            self._entries[key] = CacheEntry(value, ttl)
            self._entries.move_to_end(key)
            while len(self._entries) > self._max_size:
                self._entries.popitem(last=True)

    def get_or_load(self, key, loader, ttl_seconds=None):
        value = self.get(key)
        if value is not None:
            return value
        value = loader(key)
        self.put(key, value, ttl_seconds)
        return value

    def invalidate_prefix(self, prefix):
        with self._lock:
            for key in self._entries:
                if key.startswith(prefix):
                    del self._entries[key]

    def hit_rate(self):
        return self.hits / (self.hits + self.misses)

    def stats(self):
        with self._lock:
            return {
                "size": len(self._entries),
                "hits": self.hits,
                "misses": self.misses,
                "hit_rate": self.hit_rate(),
            }
